jax/_src/array.py

```diff
@@ -256,29 +257,58 @@ def __len__(self):
       raise TypeError("len() of unsized object") from err  # same as numpy error
 
   def __bool__(self):
-    return bool(self._value)
-
-  def __nonzero__(self):
+    if self.size == 0:
+      # deprecated 2023 September 18.
+      warnings.warn(
+        "The truth value of an empty array is ambiguous. Returning False. In the future this "
+        "will result in an error. Use `array.size > 0` to check that an array is not empty.",
+        DeprecationWarning, stacklevel=2)
+      return False
+    if self.size > 1:
+      raise ValueError("The truth value of an array with more than one element is "
+                       "ambiguous. Use a.any() or a.all()")
     return bool(self._value)
 
   def __float__(self):
+    if self.size != 1:
+      raise TypeError("Only length-1 arrays can be converted to Python scalars.")
+    if self.shape != ():
+      # Added 2023 September 18.
+      warnings.warn("Conversion of an array with ndim > 0 to a scalar is deprecated, "
+                    "and will error in future.", DeprecationWarning, stacklevel=2)
     return self._value.__float__()
 
   def __int__(self):
+    if self.size != 1:
+      raise TypeError("Only length-1 arrays can be converted to Python scalars.")
+    if self.shape != ():
+      # Added 2023 September 18.
+      warnings.warn("Conversion of an array with ndim > 0 to a scalar is deprecated, "
+                    "and will error in future.", DeprecationWarning, stacklevel=2)
     return self._value.__int__()
 
   def __complex__(self):
+    if self.size != 1:
```

Comment:
Can you pull this into a common function and just call that everywhere?

Reply:
Done